New linters and bug fixes

fangly opened the pull request.

jimhester commented on R/undesirable_function_linter.R:
```diff
@@ -0,0 +1,58 @@
+  "detach"="use roxygen2's @importFrom statement in packages, or `::` in scripts",
+  "library"= "use roxygen2's @importFrom statement in packages, or `::` in scripts",
+  "require"= "use roxygen2's @importFrom statement in packages, or `::` in scripts",
+  "loadNamespace"="use `::` or requireNamespace()",
```
I do not think `loadNamespace` should be linted, I actually prefer using it in most cases, most of the time you use it for suggested packages and do want to throw an error if a package is not installed. See http://r-pkgs.had.co.nz/namespace.html#search-path for a discussion of the differences.

fangly replied:
If you are generally aiming to follow Hadley's style and guidelines, `loadNamespace` should be left in by default.

jimhester commented on R/undesirable_function_linter.R:
```diff
@@ -0,0 +1,58 @@
+  "source"=NA,
+  "sapply"="use vapply() or lapply()",
+  "mapply"="use Map()",
+  "ifelse"="use an if () {} else {} block",
```
While it is true some (most?) of the time people use ifelse when it would be better to use `if()` there are cases where you do want to branch on a vectorized conditional, I think it is too aggressive to lint this.

fangly replied:
Ok, I'll remove it from the default. I will likely make it available a set of extended defaults for those that prefer to lint aggressively.

jimhester commented on R/undesirable_function_linter.R:
```diff
@@ -0,0 +1,58 @@
+  "sapply"="use vapply() or lapply()",
+  "mapply"="use Map()",
+  "ifelse"="use an if () {} else {} block",
+  "return"="let the last value of a function automatically be returned",
```
`return()` is still very useful for early returns, you would need to only lint this at the end of a function body.

fangly replied:
Sure, that just reflected my preference. I will remove if from the defaults. In the case of `return`, a separate linter would be preferable anyway,

jimhester commented on R/undesirable_function_linter.R:
```diff
@@ -0,0 +1,58 @@
+  "mapply"="use Map()",
+  "ifelse"="use an if () {} else {} block",
+  "return"="let the last value of a function automatically be returned",
+  "substring"="use substr()"
```
Hmm I think this is also too aggressive, substring recycles its first argument, which makes it useful for some things.

``` r
substring("abcdefghijklmnop", seq(1, 15, 5))
#> [1] "abcdefghijklmnop" "fghijklmnop"      "klmnop"
```

fangly replied:
No worries. Again, this only reflected my own personal biases.

jimhester commented on R/path_linters.R:
```diff
@@ -0,0 +1,208 @@
+  )
+}
+
+is_long_path <- function(str) {
```
I think `path` is a better name for this argument in all of these path functions.

fangly replied:
The three functions `is_valid_path()`, `is_long_path()` and `is_valid_long_path()` do indeed expect some sort of path as input. It makes sense to change the argument name as you suggested.

I am not so convinced for changing the argument name for the four functions `is_absolute_path()`, `is_root_path()`, `is_relative_path()`, `is_path()` because the input is a string and the functions determine if this is a path.

jimhester replied:
Fine, `split_path()` arguments need to be `path` and `sep` though.

fangly replied:
I thought we were splitting hairs, so I changed it to "path" anyway.
I will do the same for `split_path` as well.

jimhester commented on R/object_name_linters.R:
```diff
@@ -0,0 +1,193 @@
+#' @param style UpperCamelCase, lowerCamelCase, alllowercase, ALLUPPERCASE, snake_case or
+#'              dotted.case
+#' @export
+object_name_linter <- function(style = "lowerCamelCase") {
```
The default needs to be `snake_case`.

jimhester commented on R/comment_linters.R:
```diff
@@ -0,0 +1,169 @@
+        code_locs <<- register_code_locations(source_file, code_locs)
+      } else {
+        # non-code section: detect and report TODOs
+        if (length(file_lines) == 0L) {file_lines <- ""}  # workaround rex bug goo.gl/2rbZDy
```
I would rather have a full link rather than short url, as the full link contains information that can be used to retrieve the issue even if the link breaks in the future, such as repo username and issue number.

fangly replied:
Yes, that is understandable.

jimhester commented on R/comment_linters.R:
```diff
@@ -0,0 +1,169 @@
+#' @describeIn linters  Check that the source contains no TODO comments (case-insensitive).
+#' @param todo  Vector of strings that identify TODO comments.
+#' @export
+todo_comment_linter <- function(todo=c("todo", "fixme")) {
```
I feel like this function should just use the `getParseData()` and check for a `COMMENT` token that has text `^#[[:spaces:]]TODO`, why is all this complication needed?

fangly replied:
The function is more complicated than anticipated because not all comments are reported within `source_file$parsed_content`.

```R
# TODO 1
x <- 3 # TODO 2
foo <- function(x) { # TODO 3
  # TODO 4
  x * 2 # TODO 5
}
y <- foo(x) # TODO 6
# TODO 7
```
In the example above, only TODO 3-5 are returned as tokens. In `semicolon_terminator_linter()`, there is a similar issue that is addressed by the `locate_other_sc` function.

jimhester commented:
> Note: there is no alias for camel_case_linter(), snake_case_linter() and multiple_dots_linter() because they identify names to remove, while object_name_linter() identifies names to keep.

Yes this is true, we will probably need to keep the old definitions around until these functions are completely deprecated and gone then.

jimhester commented on R/path_linters.R:
```diff
@@ -0,0 +1,208 @@
+  for (q in type) {
+    str <- re_substitutes(str, rex(start, q, capture(zero_or_more(any)), q, end), "\\1")
+  }
+  stringi::stri_unescape_unicode(str)
```
We cannot depend on stringi, you can look at base alternatives if you need this http://stackoverflow.com/a/25310392/2055486

fangly commented:
Hi Jim,

I think I have pretty much addressed all your comments now. Good input, thanks!

The only issue left is the stringi dependency. The problem is not to escape, but to **un**escape. I am not aware of anything that does this in base. Maybe you do?

Cheers,
Florent

fangly commented:
I have implemented an unescape() function that relies on base R's `gregexpr` and `regmatches()` instead of stringi. I think there are no more open issues, Jim.

jimhester commented:
Ok I think all that is left is updating the NEWS file and README explaining the changes / listing the new linters added.

fangly commented:
This is done. It is a minor commit, so, I merging this PR now.

jimhester commented:
Please let me merge major things like this.

fangly commented:
Ok, sorry for my impatience. Thank you, Jim.
